dll/ntdll/rtl/hashtable.c

```diff
@@ -0,0 +1,1161 @@
+      CurEntry = ChainHead;
+      }
+
+    while (CurEntry->Flink != ChainHead) {
+      RTL_LIST_ENTRY *NextEntry = CurEntry->Flink;
+      RTL_DYNAMIC_HASH_TABLE_ENTRY *NextHashEntry = RtlFlinkToHashEntry(&NextEntry->Flink);
+      if (RTL_HASH_RESERVED_SIGNATURE != NextHashEntry->Signature) {
+        RemoveEntryList(&(Enumerator->HashEntry.Linkage));
+
+        NT_ASSERT(Enumerator->ChainHead != NULL);
+
+        if (Enumerator->ChainHead != ChainHead) {
+          if (IsListEmpty(Enumerator->ChainHead)) {
+            HashTable->NonEmptyBuckets--;
+            }
+
+          if (IsListEmpty(ChainHead)) {
+            HashTable->NonEmptyBuckets++;
+            }
+          }
+
+        Enumerator->BucketIndex = i;
+        Enumerator->ChainHead = ChainHead;
+
+        InitializeListHead(NextEntry, &(Enumerator->HashEntry.Linkage));
+        return NextHashEntry;
+        }
+
+      CurEntry = NextEntry;
+      }
+    }
+
+  return NULL;
+  }
+
+  /**
+   * @brief This routine reverses the effect of InitEnumeration. It decrements the
+   * NumEnumerators counter in HashTable and cleans up Enumerator state.
+   *
+   * Synchronization: The hash table lock must be held in exclusive mode.
+   *
+   * @param HashTable - Pointer to hash table on which enumerator was operating.
+   * @param Enumerator - Pointer to enumerator representing the enumeration that needs
+   * to be ended.
+   */
+VOID
+NTAPI
+RtlEndEnumerationHashTable(_In_ RTL_DYNAMIC_HASH_TABLE *HashTable,
+                                      _Inout_ RTL_DYNAMIC_HASH_TABLE_ENUMERATOR *Enumerator) {
+  DPRINT1("RtlEndEnumerationHashTable() called\n");
+  NT_ASSERT(Enumerator != NULL);
+  NT_ASSERT(HashTable->NumEnumerators > 0);
+  HashTable->NumEnumerators--;
+
+  if (!IsListEmpty(&(Enumerator->HashEntry.Linkage))) {
+    NT_ASSERT(Enumerator->ChainHead != NULL);
+
+    RemoveEntryList(&(Enumerator->HashEntry.Linkage));
+
+    if (IsListEmpty(Enumerator->ChainHead)) {
+      NT_ASSERT(HashTable->NonEmptyBuckets > 0);
+      HashTable->NonEmptyBuckets--;
+      }
+    }
+
+  Enumerator->ChainHead = FALSE;
+  }
+
+_Must_inspect_result_
+NTSYSAPI
+BOOLEAN
+NTAPI
+RtlInitWeakEnumerationHashTable(_In_ RTL_DYNAMIC_HASH_TABLE *HashTable,
+                                _Out_ RTL_DYNAMIC_HASH_TABLE_ENUMERATOR *Enumerator) {
+  DPRINT1("RtlInitWeakEnumerationHashTable() called\n");
+  return RtlInitEnumerationHashTable(HashTable, Enumerator);
+  }
+
+_Must_inspect_result_
+NTSYSAPI
+PRTL_DYNAMIC_HASH_TABLE_ENTRY
+NTAPI
+RtlWeaklyEnumerateEntryHashTable(_In_ RTL_DYNAMIC_HASH_TABLE *HashTable,
+                                 _Inout_ RTL_DYNAMIC_HASH_TABLE_ENUMERATOR *Enumerator) {
+  DPRINT1("RtlWeaklyEnumerateEntryHashTable() called\n");
+  return RtlEnumerateEntryHashTable(HashTable, Enumerator);
+  }
+
+VOID
+NTAPI
+RtlEndWeakEnumerationHashTable(_In_ RTL_DYNAMIC_HASH_TABLE *HashTable,
+                                          _Inout_ RTL_DYNAMIC_HASH_TABLE_ENUMERATOR *Enumerator) {
```

Comment:
Is this formatting that comes directly from the MS hashtable code?

Reply:
No. I get the function signatures from reactos/sdk/include/ndk/rtlfuncs.h, and created stubs with just `NOT IMPLEMENTED` note. Then, one by one, I translated functions from `msquic` to the most reasonable `Rtl` version. So, mostly those formats are from ReactOS headers. For the utility functions, I tried to keep the `msquic` version. That's why there's an inconsistency in the formatting.

I tried to apply the Code Style guidelines but there was not much details there.

Reply:
Oh, I was on mobile. You mean the parameter align issues, right? Yeah, That's probably VS code re-formatted when I saved the code before. I'll check that.

Reply:
It should be better now.